fedimint-client/src/sm/executor.rs

```diff
@@ -390,181 +405,258 @@ impl ExecutorInner {
+            Completed {
+                state: DynState,
+                outcome: ActiveOrInactiveState,
+            },
+            /// New job receiver disconnected, that can only mean termination
+            Disconnected,
         }
 
-        loop {
-            if active_state_count == 0 {
-                debug!(
-                    "No state transitions remaining, exiting execute_next_state_transitions loops"
-                );
-                return Ok(());
-            }
-            let num_states = active_state_count;
-            let num_transitions = transitions.len();
-            debug!(
-                num_states,
-                num_transitions, "Awaiting any state transition to become ready"
-            );
-            let new_state_added = async move {
-                loop {
-                    // Prioritize existing active states over new states
-                    fedimint_core::task::sleep(EXECUTOR_POLL_INTERVAL).await;
-                    let new_active_states_count = self.get_active_states().await.len();
-                    if new_active_states_count > active_state_count {
-                        return;
+        // Keeps track of things already running, so we can deduplicate, just
+        // in case.
+        let mut currently_running_sms = HashSet::<DynState>::new();
+        // All things happening in parallel go into here
+        let mut futures: FuturesUnordered<BoxFuture<'_, ExecutorLoopEvent>> =
+            FuturesUnordered::new();
+
+        re_add_sm_update_receiver(&mut futures, sm_update_rx);
+
+        // main reactor loop: wait for next thing that completed, react (possibly adding
+        // more things to `futures`)
+        while let Some(event) = futures.next().await {
+            match event {
+                ExecutorLoopEvent::New {
+                    state,
+                    sm_update_rx,
+                } => {
+                    re_add_sm_update_receiver(&mut futures, sm_update_rx);
+
+                    if currently_running_sms.contains(&state) {
+                        warn!(target: LOG_CLIENT_REACTOR, operation_id = %state.operation_id(), "Received a state machine that is already running. Ignoring");
+                        continue;
                     }
+                    let Some(meta) = self.get_active_state(&state).await else {
+                        warn!(target: LOG_CLIENT_REACTOR, operation_id = %state.operation_id(), "Couldn't look up received state machine. Ignoring.");
+                        continue;
+                    };
+
+                    let transitions = self
+                        .get_transition_for(&state, meta, &global_context_gen)
+                        .await;
+                    if transitions.is_empty() {
+                        warn!(target: LOG_CLIENT_REACTOR, operation_id = %state.operation_id(), "Received an active state that doesn't produce any transitions. Ignoring.");
```

Comment:
Shouldn't the active state be removed and made an inactive one? Otherwise this would happen on every restart.

Reply:
It still is, in `get_transition_for` (I kept it there, and only removed composing the selected here, as it seems to belong to this layer of abstraction (polling things)).